src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurDetailLineGroupForCollector.java

```diff
@@ -319,33 +320,87 @@ public void buildAndAddOriginEntries(Consumer<OriginEntryFull> entryConsumer) {
+            addOffsetOriginEntriesForUnusedAtmCashAdvanceAmountLines(temporaryCollectorBatch::addOriginEntry, unusedAtmAmountSubGroup);
         }
         
-        addOriginEntriesForCorpCardPersonalExpenseDebitLines(entryConsumer);
-        addOriginEntriesForCorpCardPersonalExpenseCreditLines(entryConsumer);
+        addOriginEntriesForCorpCardPersonalExpenseDebitLines(temporaryCorporateCardPersonalExpenseCollectorBatch::addOriginEntry);
+        addOriginEntriesForCorpCardPersonalExpenseCreditLines(temporaryCollectorBatch::addOriginEntry);
+        
+        cleanupCashAndAddCorporateCardPersonalExpenseEntries(temporaryCollectorBatch, temporaryCorporateCardPersonalExpenseCollectorBatch);
+        
+        addEntries(entryConsumer, temporaryCollectorBatch.getOriginEntries());
+    }
+    
+	private void cleanupCashAndAddCorporateCardPersonalExpenseEntries(CollectorBatch temporaryCollectorBatch,
+			CollectorBatch temporaryCorporateCardPersonalExpenseCollectorBatch) {
+
+		List<OriginEntryFull> paymentOffsetCashEntries = temporaryCollectorBatch.getOriginEntries().stream().filter(
+				entry -> collectorHelper.paymentOffsetObjectCode.equalsIgnoreCase(entry.getFinancialObjectCode()))
+				.collect(Collectors.toList());
+		KualiDecimal cashTotal = calculateTotalAmountForEntries(paymentOffsetCashEntries);
+
+		List<OriginEntryFull> paymentOffsetCorporateCardPersonalExpenseDebitEntries = temporaryCorporateCardPersonalExpenseCollectorBatch
+				.getOriginEntries().stream().filter(entry -> collectorHelper.paymentOffsetObjectCode.equalsIgnoreCase(entry.getFinancialObjectCode()))
+				.collect(Collectors.toList());
+		List<OriginEntryFull> nonPaymentOffsetCorporateCardPersonalExpenseDebitEntries = temporaryCorporateCardPersonalExpenseCollectorBatch
+				.getOriginEntries().stream().filter(entry -> !collectorHelper.paymentOffsetObjectCode.equalsIgnoreCase(entry.getFinancialObjectCode()))
+				.collect(Collectors.toList());
+		
+		KualiDecimal corporateCardPersonalExpenseDebitTotal = calculateTotalAmountForEntries(
+				paymentOffsetCorporateCardPersonalExpenseDebitEntries);
+
+		if (cashTotal.isGreaterThan(corporateCardPersonalExpenseDebitTotal)) {
+			KualiDecimal totalDeductionsLeft = corporateCardPersonalExpenseDebitTotal;
+			for (OriginEntryFull entry : paymentOffsetCashEntries) {
+				KualiDecimal newTransactionAmount = entry.getTransactionLedgerEntryAmount().subtract(totalDeductionsLeft);
+				if (newTransactionAmount.isNegative()) {
+					totalDeductionsLeft = newTransactionAmount;
+					newTransactionAmount = KualiDecimal.ZERO;
+				} else {
+					totalDeductionsLeft = KualiDecimal.ZERO;
+				}
+				entry.setTransactionLedgerEntryAmount(newTransactionAmount);
+			}
+		} else {
+			addEntries(temporaryCollectorBatch::addOriginEntry, paymentOffsetCorporateCardPersonalExpenseDebitEntries);
+		}
+		addEntries(temporaryCollectorBatch::addOriginEntry, nonPaymentOffsetCorporateCardPersonalExpenseDebitEntries);

```

Comment:
This whole section has the potential to reorder some of the origin entries, so if further code changes are made, the unit tests may need an adjustment if they have order-dependent expectations for the generated entries.